driver/compile.go

```diff
@@ -357,3 +360,118 @@ func computeColumnsR(p ast.Proc, colset map[string]struct{}) (map[string]struct{
 		panic("proc type not handled")
 	}
 }
+
+func copyProcs(ps []ast.Proc) []ast.Proc {
+	var copies []ast.Proc
+	for _, p := range ps {
+		b, err := json.Marshal(p)
+		if err != nil {
+			panic(err)
+		}
+		proc, err := ast.UnpackJSON(nil, b)
+		if err != nil {
+			panic(err)
+		}
+		copies = append(copies, proc)
+	}
+	return copies
+}
+
+func buildSplitFlowgraph(branch, tail []ast.Proc, mergeField string, reverse bool, N int) *ast.SequentialProc {
+	pp := &ast.ParallelProc{
+		Node:              ast.Node{"ParallelProc"},
+		Procs:             []ast.Proc{},
+		MergeOrderField:   mergeField,
+		MergeOrderReverse: reverse,
+	}
+	for i := 0; i < N; i++ {
+		pp.Procs = append(pp.Procs, &ast.SequentialProc{
+			Node:  ast.Node{"SequentialProc"},
+			Procs: copyProcs(branch),
+		})
+	}
+	return &ast.SequentialProc{
+		Node:  ast.Node{"SequentialProc"},
+		Procs: append([]ast.Proc{pp}, tail...),
+	}
+}
+
+func decomposable(rs []ast.Reducer) bool {
+	for _, r := range rs {
+		cr, err := rcompile.Compile(r)
+		if err != nil {
+			return false
+		}
+		if _, ok := cr.Instantiate().(reducer.Decomposable); !ok {
+			return false
+		}
+	}
+	return true
+}
+
+// parallelizeFlowgraph takes a sequential proc AST and tries to
+// parallelize it by splitting as much as possible of the sequence into
+// N parallel branches. The boolean return argument indicates whether
+// distribution succeeded.
+func parallelizeFlowgraph(seq *ast.SequentialProc, N int, inputSortField string, inputSortReversed bool) (*ast.SequentialProc, bool) {
+	for i := range seq.Procs {
+		switch p := seq.Procs[i].(type) {
+		case *ast.CutProc, *ast.FilterProc, *ast.PassProc, *ast.PutProc, *ast.RenameProc:
+			// Stateless procs: continue until we reach one of the procs below at
+			// which point we'll either split the flowgraph or see we can't and return it as-is.
+			continue
```

Comment:
@henridf, @alfred-landrum: I don't think this does the right thing if cut, put, or rename alters the sort field. Is that handled elsewhere?

Reply:
Good catch @nwt. Addressed in #1171